### DISCUSSION TITLE
`Ctrl+R` Immediate command execution – What do you think?

### DISCUSSION THREAD

**Discussion by jerrygreen** [fish-shell/fish-shell#9938] [Ideas]
There's this reverse-i-search feature which lets you search through your command history:

![gif](https://github.com/fish-shell/fish-shell/assets/13215662/ed9af2d8-97e7-4b16-ac27-1a528de964d6)

- In bash, it simply iterates between commands each time you press `Ctrl+R`
- In fish, you only press `Ctrl+R` once and then select an item from search with arrow-keys

In comparison to bash, this feature is much more rich in fish, showing multiple results at once.

Convenient. But what makes it less convenient at the same time is this:

- Now you have to press `Enter` twice

This can be explained simply:

1. With your first `Enter` you first CHOOSE what a particular search result
2. With second `Enter` you then EXECUTE the command

So in-between of those two you have a chance to change the original command, thus the necessity of such.

Not sure about everyone but I personally TOO OFTEN just want to re-use exact command I used. No changing at all.

|  IT WOULD BE MORE CONVENIENT IF ONLY ONE `ENTER` WOULD BE REQUIRED, – JUST LIKE IN BASH. |
| --- |

Thus, I ask you if you agree with this statement. And if yes, what other hotkey do you think would be best to replace existing `Enter` functionality? So you still would be able to somehow edit the selected command, while `Enter` would just execute the command right away just like in bash.

> This topic basically came from here:
> - https://github.com/fish-shell/fish-shell/issues/9934

**Comment by faho** [comment]
> IT WOULD BE MORE CONVENIENT IF ONLY ONE ENTER WOULD BE REQUIRED

My immediate reaction to this proposal is a fairly clear "NO".

I do not wish to make it harder to edit the line and I don't want to change this *now* so people used to the current behavior accidentally execute something they didn't mean to.

And pressing enter twice is *extremely* easy. Your finger is already on the key.

And if the first enter didn't allow you to edit the command, there wouldn't be an intuitive way to do so! You can't press space (that inserts a space), you can't press escape (that exits the pager), you would have to press a special key. And that makes the feature quite a bit worse.

What's more:

> , – JUST LIKE IN BASH.

Fish is not bash, and I would appreciate it if we could diverge in the minute key strokes, instead of being measured at bash-compatibility at the key-for-key level.

In #9934 you even refer to it as "In normal bash", and I want us to get away from that perception. Bash isn't normal.

**Reply by jerrygreen** [reply]
> "In normal bash", and I want us to get away from that perception. Bash isn't normal

Bash and zsh were the best thing to me before I found fish, so... It is okay. Compare this to some windows `cmd` shell with their bat scripts or whatever they're using in their powershell... Wheeesh.

> And pressing enter twice is extremely easy

Working with mouse is also extremely easy. But most console-people like you and I prefer to use hotkeys as more efficient.

> And if the first enter didn't allow you to edit the command, there wouldn't be an intuitive way to do so!
> You can't press space ... you can't press escape ... you would have to press a special key

Yeah that's the main problem, basically the reason for this discussion. I thought maybe people come up with some hotkey idea that already exists somewhere for similar thing or something. I'm still relatively new with fish so idk.

For example, what does consecutive `Ctrl+R` do? In bash it's basically the way of navigating search results. But in fish you navigate with arrows, while consecutive `Ctrl+R` doesn't seem to do anything... Maybe the solution would be to make it like a toggle: pressing it first enables the search, and second `Ctrl+R` disables it (while preserving the selected text for further edits).

> and I don't want to change this _now_ so people used to the current behavior accidentally execute something they didn't mean to

That's fair point. But there's two: fish users and newcomers (many of those are from bash or zsh):

- For newcomers: once they see multiple items in search they at first see minor yet cool convenience. But after that they see command not executed and face minor yet annoying... Annoyance. So it's not just 100% improvement. It's 50/50 improvement. A compromise.
- For fish users, it might be best to make it a feature flag for now, which can be enabled in config early, and then make it the default in next major version as being considered «breaking change». So people can be ready for this change. Read about it in patch notes, etc. (I might have said this too early but please continue reading)

For you in particular, as one of the first fish users basically, you might be very against this idea as you're already having a habit of how it always was in fish. Understandable. But if you try this «one-enter» approach and take some time to get used to it, after that you will most probably love the concept. The selected items will turn out to be kind of like... Buttons to select. Imagine if in browsers you had to press mouse button twice? Or if after selecting a link with hotkeys you'd have to press enter twice? Nonsense. And those two kinds of software aren't something diagonally-different:

1. You can tab between elements in browser
2. You can tab between search results in fish (tab and shift+tab already supported in fish, – just like in any web browser)

Choosing newcomers side is _kind-of-compromise_ over the habits of fish users but I think it's worth it. Imagine what might be, say, in 10 years? Imagine if Apple would make fish their default shell in new macOS 2028 for example. Userbase might as well quadruple. Does this really worth sticking to a bad habit of double-pressing, while most of these new users would find this annoyance of reduced efficiency? Even so slightly. Just another something to take into consideration.

Think about it. I genuinely believe that you will love this one-enter idea once get used to it. It doesn't make fish too similar to bash, nor ugly. On the inverse: it is a sensical thing to do.

(If you eventually agree, just ping me here, I'll convert the gist of it into a humanely-readable issue, with this consecutive `Ctrl+R` toggle)

**Reply by faho** [reply]
>But in fish you navigate with arrows, while consecutive Ctrl+R doesn't seem to do anything...

It opens the next page of search results. If none exists, it flashes the commandline to indicate it doesn't.

>Working with mouse is also extremely easy.

Moving the hand to the mouse and pressing the key you are already on twice are two very different things.

>Imagine if Apple would make fish their default shell in new macOS 2028 for example

This is an entirely baseless idea.

And the new users wouldn't rebel because they need to disambiguate in one specific case.

(they would rebel for other reasons, which is why Apple would probably never do this)

> this annoyance of reduced efficiency?

Mate it's a single key press. We're not asking you to fill out a form, we're asking you to choose between "edit this thing more" (which is something that you might do - I know I do regularly) and "run this thing now", by either editing or pressing the "run this thing" button again.

>I genuinely believe that you will love this one-enter idea once get used to it.

Trust me I won't.

It's also useful to know, for context, that this is how the pager behaves in general - when you do `git <TAB>` and press enter, that will select the completion and not execute - because you don't want to just run `git add` without another argument.

So we have some nice consistency there.


----

I would like you to step back a bit and think about what you are asking, and then argue based on that.

You're being a bit hyperbolic here ("what if Apple", "using a mouse", "reduced efficiency"), and it helps to keep to the idea at hand.

You want to change how a specific action works in a specific case, by making another action harder. That's fundamentally a trade-off, and one I don't think you are winning.

Neither of us has found a good solution to allowing the other action, *intuitively*, while the current solution allows both possible actions, *intuitively*. That makes the current way not terrible already!

**Comment by krobelus** [comment]
I like Ctrl+R to toggle but I fear that it's not discoverable. It would be unfortunate if a significant percentage of users didn't know how to edit the search result. It's a shame we can't use Tab for switching between input fields.
We can definitely allow rebinding keys to make this possible

**Comment by piersy** [comment]
@jerrygreen I'm using the following binding to be able to insert and execute in one go from the history pager

```
bind ctrl-n 'commandline --function execute; commandline --function execute'
```

I tried using a conditional function to bind enter to the above while the pager was active, but unfortunately when executing inside a function the first `execute` call exits the keybinding context and the second execute call is lost. So I think this is the best you can get right now.

I would be in favour of adding a new command to accept_and_execute the contents of the pager, then people could bind that to what they like, and achieve what you were originally looking for.
